File: typescript/src/stream-usage.ts
/** Parsed SSE `event: usage` payload after a chat/agent stream completes. */
export type StreamUsageEvent = {
  runCredits: number;
  threadCreditsTotal: number;
  balanceRemaining?: number;
  executionId?: string | null;
  breakdown?: {
    ai_model: number;
    tool: number;
    sandbox: number;
    browser: number;
    workflow: number;
    media: number;
  };
};

export type ParsedSseEvent = {
  event: string;
  data: unknown;
  id?: string;
};

/**
 * Parse a raw SSE chunk (one or more events) into structured events.
 * Works with fetch() streams and Node Readable streams.
 */
export function parseSseChunk(buffer: string): { events: ParsedSseEvent[]; remainder: string } {
  const events: ParsedSseEvent[] = [];
  const blocks = buffer.split('\n\n');
  const remainder = blocks.pop() ?? '';

  for (const block of blocks) {
    const lines = block.split('\n');
    let event = 'message';
    let id: string | undefined;
    const dataLines: string[] = [];

    for (const line of lines) {
      if (line.startsWith('event:')) {
        event = line.slice(6).trim();
      } else if (line.startsWith('data:')) {
        dataLines.push(line.slice(5).trimStart());
      } else if (line.startsWith('id:')) {
        id = line.slice(3).trim();
      }
    }

    if (!dataLines.length) continue;
    const raw = dataLines.join('\n');
    let data: unknown = raw;
    try {
      data = JSON.parse(raw);
    } catch {
      // keep raw string
    }
    events.push({ event, data, id });
  }

  return { events, remainder };
}

/** Return the last `usage` event from a list of parsed SSE events. */
export function extractStreamUsage(events: ParsedSseEvent[]): StreamUsageEvent | null {
  for (let i = events.length - 1; i >= 0; i -= 1) {
    if (events[i]?.event === 'usage') {
      return events[i].data as StreamUsageEvent;
    }
  }
  return null;
}

File: python/src/flowra/stream_usage.py
"""SSE parsing helpers for chat/agent stream usage events."""

from __future__ import annotations

import json
from typing import Any, Optional, TypedDict


class StreamUsageBreakdown(TypedDict, total=False):
    ai_model: float
    tool: float
    sandbox: float
    browser: float
    workflow: float
    media: float


class StreamUsageEvent(TypedDict, total=False):
    runCredits: float
    threadCreditsTotal: float
    balanceRemaining: float
    executionId: Optional[str]
    breakdown: StreamUsageBreakdown


class ParsedSseEvent(TypedDict, total=False):
    event: str
    data: Any
    id: str


def parse_sse_chunk(buffer: str) -> tuple[list[ParsedSseEvent], str]:
    """Parse a raw SSE chunk into structured events.

    Returns ``(events, remainder)`` where ``remainder`` is an incomplete
    trailing block that should be prepended to the next chunk.
    """
    events: list[ParsedSseEvent] = []
    blocks = buffer.split("\n\n")
    remainder = blocks.pop() if blocks else ""

    for block in blocks:
        lines = block.split("\n")
        event = "message"
        event_id: Optional[str] = None
        data_lines: list[str] = []

        for line in lines:
            if line.startswith("event:"):
                event = line[6:].strip()
            elif line.startswith("data:"):
                data_lines.append(line[5:].lstrip())
            elif line.startswith("id:"):
                event_id = line[3:].strip()

        if not data_lines:
            continue

        raw = "\n".join(data_lines)
        data: Any = raw
        try:
            data = json.loads(raw)
        except json.JSONDecodeError:
            pass

        parsed: ParsedSseEvent = {"event": event, "data": data}
        if event_id is not None:
            parsed["id"] = event_id
        events.append(parsed)

    return events, remainder


def extract_stream_usage(events: list[ParsedSseEvent]) -> Optional[StreamUsageEvent]:
    """Return the last ``usage`` event from a list of parsed SSE events."""
    for item in reversed(events):
        if item.get("event") == "usage":
            data = item.get("data")
            if isinstance(data, dict):
                return data  # type: ignore[return-value]
    return None
